rfcs/20230418-add-pow-graph-op/README.md

```diff
@@ -0,0 +1,267 @@
+   transforming.
+
+### Option 4: Add Both Binary Element-wise Pow and Rsqrt
+
+This option will add two operation in Graph API.
+`Pow` takes two input tensors `src` and `exponent`.
+`Rsqrt` is added as an eltwise operation.
+
+| Pow       | Argument Name    | Required or Optional       | Data Type     |
+|-----------|------------------|----------------------------|---------------|
+| input     | `src`            | Required                   | f32,f16,bf16* |
+| input     | `exponent`       | optional                   | f32,f16,bf16  |
+| attribute | `auto_broadcast` | optional, default: 'numpy' | string        |
+| output    | `dst`            | Required                   | f32,f16,bf16* |
+
+*: input and output must have the same data type.
+
+| Rsqrt  | Argument Name | Required or Optional | Data Type     |
+|--------|---------------|----------------------|---------------|
+| input  | `src`         | Required             | f32,f16,bf16* |
+| output | `dst`         | Required             | f32,f16,bf16* |
+
+*: input and output must have the same data type.
+
+Pros.
+
+1. clearly mapping semantics of `Pow` and `Rsqrt` for frameworks.
+
+Cons.
+
+1. need to add two new operations: `Pow` and `Rsqrt`
+2. `Pow` cannot be optimized with oneDNN primitive.
+
+### Option 5: Add Binary Element-wise Pow and Use Reciprocal and Sqrt as Rsqrt
+
+The option will add only one operation `Pow` and use `Reciprocal` and `Sqrt` to
+cover the funcionality of `Rsqrt`.
+`Pow` takes two input tensors `src` and `exponent`.
+`Rsqrt` reuse `Reciprocal` and `Sqrt`: Rsqrt(src) = Reciprocal(Sqrt(src)).
+`Reciprocal` and `Sqrt` are already supported in oneDNN Graph API.
+
+| Pow       | Argument Name    | Required or Optional       | Data Type     |
+|-----------|------------------|----------------------------|---------------|
+| input     | `src`            | Required                   | f32,f16,bf16* |
+| input     | `exponent`       | optional                   | f32,f16,bf16  |
+| attribute | `auto_broadcast` | optional, default: 'numpy' | string        |
+| output    | `dst`            | Required                   | f32,f16,bf16* |
+
+Pros.
+
+1. `Pow` is compatible with mainstream frameworks.
+2. `Rsqrt` reuse current op set. Will keep the op set small.
+
+Cons.
+
+1. framework bridge will need to split `Rsqrt` to two oneDNN Graph operations.
+2. `Pow` cannot be optimized with oneDNN primitive.
+
+## Conclusion
+In the technical meeting on ww17.2, option 1 is chosed as current solution:
+unary eltwise Pow with attribute, Rsqrt reused Pow with attribute -1/2
+
+1. Easy for framework to integrate
+2. Can leverage oneDNN primitive
+3. Don’t have runtime exponent Pow in current models
+
+For runtime exponent, need revisit this proposal in the future. Maybe add a new
+operation dynamicPow.
+### API changes
+
+New operation(s) will be added in Graph API.
+
+### Backend changes
+
+Graph Compiler backend will fuse new added operation(s) into MLP pattern for LLM
```

Comment:
@TaoLv @rongzha1 Will primitive-based backend support this op? Eltwise primitive does support pow algorithm.

Reply:
primitive-based backend will be added on demand. Currently no single Pow optimization request from frameworks.
Or can we target it on v3.3?

Reply:
Yes, once the op is added, we will support the post-ops fusion based on the new op as usual, eg. conv/matmul + pow patterns. But so far we don't have the plan to support the RMSNorm fusion mentioned here.